### DISCUSSION TITLE
Pulling data

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#439] [General]
I'm not sure if I've asked this before, I can't find any record so here it is.

1. Is there anyway to pull ALL the tables, other than to get the table reference and call `PullItemsAsync`
2. Is there anyway to get a result of what was pulled ? It would be good if I could call `PullItemsAsync` and get a result.

I know I could implement this but wanted to double check there is nothing already there or in the pipeline.

**Comment by adrianhall** [comment]
``` csharp
await client.PullItemsAsync();
```

What do you mean by "get a result" precisely?

**Reply by henda79** [reply]
Its open to ideas, but at the moment, I make the call and have no idea whether I got 0, 1 or 100 rows back.

I think there needs to be a way we can see what we just pulled, whether its a list of the rows, or maybe just Id's

**Reply by henda79** [reply]
regarding your code snippet, I don't have that method in the client.

I have a fork of this project which is up to date and I use that code in my project.

**Reply by adrianhall** [reply]
Ah - yes - I forgot I didn't implement that.  I did the `client.PushItemsAsync()` so you could push all tables in advance.  Feel free to add an issue for the `PullItemsAsync()`.

For the "what info", how about I define an event handler on the client.  When a push or pull happens, you can get an event for:

* Push started on table
* Push finished on table with stats
* Pull started on table
* Pull finished on table with stats

Where the event type would be something like:

``` csharp
struct SynchronizationEventArgs {
  public string TableName { get; }
  public SynchronizationEventType EventType {get; } // Push, Pull, Purge
  public string OperationId {get; } // Specified in PushOptions, PullOptions, PurgeOptions, or auto-generated.
  public long ItemsSuccessfullyProcessed {get; }
  public long ItemsRejected {get; }
}
```

I could also provide an event handler on the store for individual items (stored, deleted, etc.)

Thoughts?

**Reply by henda79** [reply]
Yes, something like that. I was thinking about have the method PullItemsAsync return Task<SomeSyncData> rather than just Task, but having an event handler could be a better plan so we can hook into where ever we need to.

**Reply by adrianhall** [reply]
File an issue for that as well - I'll get to it when I next sit down to work on the .NET client

**Reply by henda79** [reply]
If you don't mind, I'll have a go at implementing the PullTablesAsync myself, nothing fancy just maybe 2 methods to allow for pull all or pull a selection.

**Reply by adrianhall** [reply]
File the issue first :-)

For the implementation, probably the best way of doing it is to add a method to the DeltaTokenStore that retrieves the list of tables.  Then you can implement PullAllTablesAsync(...) on the SyncContext to do a PullTableAsync on each one.  Think about the configuration - maybe an exclude list (for the case where I want to pull all tables except for one).  Once you have the SyncContext implementation, wiring it into the client is trivial.

**Reply by henda79** [reply]
ok, will do. I've just spent the last 15 minutes trying to work out where best to get the table list from. thanks for the pointer. I'll file an issue.

**Reply by henda79** [reply]
I've had a quick look at the DeltaTokenStore but I can't see any way to get a list of the tables.

I can see there is a list in the OfflineStore but only in the concrete class OfflineSQLiteStore, private field tableMap

Am I missing something ?

**Reply by adrianhall** [reply]
`DeltaTokenStore` has a reference to `OfflineStore`, which is an `IOfflineStore`.  `IOfflineStore` has a `GetPageAsync()` which returns a page of items according to the query provided.  There is also a `FuncAsyncEnumerable` which takes query and can convert a `Page<T>` into an `IAsyncEnumerable<T>`, generally used like this:

``` csharp
public IAsyncEnumerable<T> GetAsyncItems<T>(somethign query) 
{
  return new FuncAsyncPageable<T>(nextLink => GetNextPageAsync(query, nextLink);
}
```

Then you implement `GetNextPageAsync(query, nextLink)` like this:

``` csharp
public async Task<Page<T>> GetNextPageAsync<T>(something query, string nextLink, CancellationToken token = default) 
{
    var page = // Do whtever here to get a Page<T>
    return page;
}
```

Check out `OfflineTable.generic.cs` for a concrete example.  Now you have `GetAsyncItems()`, you can do

``` csharp
var tables = await GetAsyncItems().ToListAsync();
```

To get a `List<T>` of the items you return.  Some follow-up - T is probably DeltaTokenInfo - a new struct that has a table name and the DateTimeOffset.  You now have all the elements required to implement a public `Task<List<DeltaTokenInfo>> GetDeltaTokensAsync(CancellationToken token = default)` on the DeltaTokenStore.

**Reply by henda79** [reply]
Maybe this is a one for you, I'm not able to convert `Page<JObject>` to `Page<T>` as there is no access to the `ServiceSerializer` in the DeltaTokenStore or through the OfflineStore in the implemented GetNextPageAsync method.

```
        public IAsyncEnumerable<T> GetAsyncItems<T>(QueryDescription query) 
        {
            return new FuncAsyncPageable<T>(nextLink => GetNextPageAsync<T>(query, nextLink));
        }

        private async Task<Page<T>> GetNextPageAsync<T>(QueryDescription query, string nextLink, CancellationToken cancellationToken = default)
        {
            var page = await OfflineStore.GetPageAsync(query, cancellationToken);

            return new Page<T>
            {
                Count = page.Count,
                Items = page.Items.Select(m => Serializer.Deserialize<T>(m)).ToArray(),
                NextLink = page.NextLink
            };
        }
```

does this work 

```
            return new Page<T>
            {
                Count = page.Count,
                Items = page.Items.Select(m => m.ToObject<T>()).ToArray(),
                NextLink = page.NextLink
            };
```

or should the Serilializer be passed in somehow ?

Also, what is the query ? It seems it is type of QueryDescription but this requires a table name in the constructor.

Sorry, I'm totally lost.

**Reply by henda79** [reply]
Anyway, I don't think this approach will work.

Imagine you have a fresh client with no pulled data and we are not defining all the tables upfront (which is now allowed), using this `PullAllTablesAsync` method would do nothing. It would only work if the tables are declared upfront by using `DefineTable<>`

We would need to get the tables from the server first...

**Reply by adrianhall** [reply]
Good point - Probably best to leave it with me.  To get a list of defined tables, I'm going to have to do some infrastructure changes in the AbstractOfflineStore and the SQLiteStore.  Since I'm delving into that area for the delta-store, I'll do this at the same time.
